### PR TITLE
Merge branch palmer_penguins to complete 3 milestone.

### LINKED ISSUE #4: Palmer Penguins Dataset - Warning Handling and Documentation
## ✅ Excellent Work
* **Complete requirements**: All 4 split ratios, one-hot encoding, stratified splits implemented
* **Advanced preprocessing**: IterativeImputer usage shows good understanding
* **Multi-class handling**: Proper 3-class stratified splits
* **Professional code**: Both modular and direct implementation versions
* **Outstanding results**: High accuracy (95-100%) demonstrates good dataset choice

### 1. Warning Suppression (+0.3 points)
Fix FutureWarnings from pandas and seaborn:

```python
import warnings
warnings.filterwarnings('ignore', category=FutureWarning)
```

### 2. Feature Documentation (+0.4 points)
Add feature descriptions:

```markdown
### Feature Descriptions:
- bill_length_mm: Length of penguin's bill in millimeters
- bill_depth_mm: Depth of penguin's bill in millimeters
- flipper_length_mm: Length of penguin's flipper in millimeters
- body_mass_g: Body mass in grams
- sex_male: Gender (True=Male, False=Female)
- island_Dream, island_Torgersen: Island location (one-hot encoded)
```

### 3. Class Balance Analysis (+0.3 points)
Add imbalance ratio:

```python
print(f"Class imbalance ratio: {y.value_counts().max()}/{y.value_counts().min()} = {y.value_counts().max()/y.value_counts().min():.2f}")
```

## **Current Score: 9.3/10 ⭐**

**Special note**: Using IterativeImputer instead of SimpleImputer shows excellent preprocessing knowledge!

### PR DESCRIPTION
**Pull request by OriginalNVK**
- Fixes #4 
- Complete visualize decision tree.
- Complete analyze accuracy with tree depth

### CODE REVIEW

File: palmer_penguins_analysis.py
Comment (eNKay2408) [inline palmer_penguins_analysis.py:1]: Why don't you put all .py files in /src folder?
Such as: debug.py, palmer_penguins_analysis.py, run_notebook.py
I think we should not put them in the root directory because it is difficult to manage

File: notebooks/02_palmer_penguins_analysis.ipynb [hunk 1]
```diff
@@ -146,16 +330,155 @@
...
+   "execution_count": 7,
    "id": "aa834647",
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "C:\\Users\\khanh\\AppData\\Local\\Temp\\ipykernel_8620\\2625201507.py:6: FutureWarning: \n",
```
Comment (eNKay2408) [inline notebooks/02_palmer_penguins_analysis.ipynb:353]: 💡 Minor: Quick fix for seaborn warning - add hue parameter:
`sns.countplot(data=df, x='species', hue='species', legend=False, palette="viridis")`

File: notebooks/02_palmer_penguins_analysis.ipynb [hunk 2]
```diff
@@ -68,22 +106,140 @@
...
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "d:\\Đại Học\\Kì 2 - Năm 3\\Cơ Sở AI\\Đồ Án Thực Hành\\Decision-Tree\\AI-Decision-Tree\\notebooks\\..\\src\\preprocessing.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.\n",
+      "The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.\n",
+      "\n",
+      "For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.\n",
```
Comment (eNKay2408) [inline notebooks/02_palmer_penguins_analysis.ipynb:183]: 💡 Minor: Use pandas-recommended approach to avoid deprecation:
`df = df.assign(sex=df['sex'].fillna(df['sex'].mode()[0]))`

File: palmer_penguins_analysis.py [hunk 3]
```diff
@@ -52,4 +52,33 @@
 
 print("Preprocessing completed!")
 
+# Train test split
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
+print(f"\nTrain set: {X_train.shape[0]} samples, Test set: {X_test.shape[0]} samples")
 
+# Train decision tree
+print("\nTraining decision tree...")
+model = DecisionTreeClassifier(criterion='entropy', random_state=42)
```
Comment (eNKay2408) [inline palmer_penguins_analysis.py:61]: ✅ Good: Using information gain (entropy) as required

File: src/preprocessing.py [hunk 4]
```diff
@@ -1,20 +1,70 @@
...
+        
+    try:
+        # Đọc dữ liệu
+        df = pd.read_csv(path)
+        print(f"Đã đọc được file với {df.shape[0]} dòng và {df.shape[1]} cột.")
+        
+        # Impute numeric columns
+        numeric_cols = df.select_dtypes(include='number').columns
+        print(f"Các cột số: {list(numeric_cols)}")
+        imputer = IterativeImputer(max_iter=10, random_state=42)
```
Comment (eNKay2408) [inline src/preprocessing.py:37]: ✅ Good choice: Using IterativeImputer for handling missing values shows advanced preprocessing

File: palmer_penguins_analysis.py [hunk 5]
```diff
@@ -52,4 +52,33 @@
 
 print("Preprocessing completed!")
 
+# Train test split
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
```
Comment (eNKay2408) [inline palmer_penguins_analysis.py:56]: ✅ Good: Stratified splitting maintains class distribution across all train/test ratios

File: notebooks/02_palmer_penguins_analysis.ipynb [hunk 6]
```diff
@@ -198,10 +536,141 @@
...
+      "   Chinstrap       1.00      1.00      1.00        14\n",
+      "      Gentoo       0.96      1.00      0.98        25\n",
+      "\n",
+      "    accuracy                           0.99        69\n",
+      "   macro avg       0.99      0.99      0.99        69\n",
+      "weighted avg       0.99      0.99      0.99        69\n",
+      "\n",
+      "\n",
+      "Evaluating model trained on split 4 (90-10):\n",
+      "Test Accuracy: 1.0000\n"
```
Comment (eNKay2408) [inline notebooks/02_palmer_penguins_analysis.ipynb:642]: ✅ Impressive: Perfect accuracy on 90/10 split shows excellent model performance

**Review by eNKay2408 — APPROVED** [review]
🎉 Outstanding implementation! Exceeds all assignment requirements with:
- Perfect task completion (2.2, 2.3, 2.4)
- Excellent accuracy results (95-100% across all splits)
- Professional-quality visualizations and analysis
- Strong biological insights about penguin species classification
- Well-structured code with proper documentation

The depth analysis (optimal at 4+) and flipper length as main feature show strong grasp of ML and domain knowledge

Minor suggestions are just for code hygiene - current implementation fully satisfies all milestone requirements
